Add html/markdown translators.

luciansmith opened the pull request:
Add the ability to set and get SBase notes and Constraint messages in the 'markdown' format, as translated md->html using maddy, and html->md using html2md.

I made no attempt to save the raw markdown for now; the functions simply translate to and from each other.  If the user wants to store the raw markdown as an annotation, they'll need to do that themselves (at least for now).

Function names are

setNotesFromMarkdown
setMessageFromMarkdown

getNotesMarkdown
getMessageMarkdown

There wasn't a 'setNotesString' function, hence the 'From'.  But the names are negotiable if someone thinks they should be changed.

From #416.

luciansmith committed:
Add html/markdown translators.

Add the ability to set and get SBase notes and Constraint messages in the 'markdown' format, as translated md->html using maddy, and html->md using html2md.

luciansmith committed:
Wrap with EXTERN.

luciansmith commented:
OK, so maddy (at least) requires C++11, which we are not set up to handle.  Should we?  It's 15 years old; it seems safe...

luciansmith commented:
Update: it's just some of our github actions that don't use C++11.  In general, the other actions compile and run fine.  So if we can update the github actions, we'd be OK?

luciansmith committed:
Fix link parser for maddy; add test.

Links that used title text were being translated to markdown by html2md, but couldn't be parsed back by maddy.  Fixed here, and also submitted a PR to maddy directly.

fbergmann commented on src/sbml/maddy/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+# This project is licensed under the MIT license. For more information see the
+# LICENSE file.
+
+cmake_minimum_required(VERSION 3.25)
```
since maddy requries cmake 3.25 it would be best that all the other requirements are also upgraded currently we still have a wider range that we allow compilation with cmake_minimum_required(VERSION 2.8.4...3.19). We might also have to upgrade the c++ standard to c++14 which is fine with me.

luciansmith replied:
I included the maddy CMakeLists file just because it was in the original, but we don't actually use it.  If this is confusing, I could just remove the file entirely?

fbergmann replied:
ok, as long as we dont require it that is fine then. Might make sense to remove it, but in any case that does not matter then. Thanks!

fbergmann reviewed (changes requested):
we need to upgrade the cmake build to require c++14 everywhere at least and a more recent cmake version.

fbergmann commented:
> Update: it's just some of our github actions that don't use C++11. In general, the other actions compile and run fine. So if we can update the github actions, we'd be OK?

yes we ought to update the actions and the cmake scripts to ensure that everyone is made aware.

fbergmann commented:
@luciansmith i actually would prefer if we had a general static util methods something like `util_html_to_markdown` / `util_markdown_to_html`. With c overloads. That would be useful from all other projects using libsbml. We'd then be just calling those methods everywhere. This would reduce the places where we need the new includes, and would have the concatenation also happening only once. Thanks!

luciansmith committed:
Updates as requested from @fbergmann

* Both libraries now wrapped in util_ functions.
* Update github actions to C++11 instead of 98.

luciansmith committed:
Change html2md to take const string ref instead of string ref.

The compilers complained about taking temp output as input when it wasn't const.  Will need to submit PR to html2md as well.

luciansmith committed:
Add C wrappers.

luciansmith committed:
Free strings in test.

luciansmith commented:
It looks like you can't actually set a minimum C++ version in CMake; you can only set a particular C++ version ('use C++ 11 always'), unless you want to check the flags manually ('make sure the user has '-std=c++11' or '-std=c++14' or ...'). See https://stackoverflow.com/questions/48148275/inducing-minimal-c-standard-version-in-cmake

So for now, I just changed the github action to use 11 instead of 98.  (We didn't need C++14 yet, just 11.)  Is that reasonable for now?

In other news, I added the functions you requested, but I wasn't sure about the names.  For now, I went with

util_html_to_markdown
util_markdown_to_html
util_html_to_markdown_c
util_markdown_to_html_c

The latter two feel clunky to me, but I couldn't think of anything better to name the C versions of those functions.  Happy to take other suggestions, if you have them.

luciansmith committed:
Test the target_compile_features of CMake.

New library needs C++; try to tell CMake this.

luciansmith committed:
Revert store-artefact to 98 as well.

CMake now should add c++11 as needed to the libraries.

luciansmith committed:
Try to see what flags are actually set when running.

luciansmith committed:
Fix format.

luciansmith committed:
Fix again?

luciansmith committed:
Merge pull request #422 from sbmlteam/test-target_compile_features

Test target compile features

luciansmith commented:
I was wrong!  You can indeed tell CMake to look for features you need with target_compile_features.  Added and tested.  It actually works even with the original github actions that try to use c++98; it adds the c++11 flag automatically.

luciansmith committed:
Revert header changes for cleaner merge.

luciansmith committed:
Merge pull request #423 from sbmlteam/test-target_compile_features

Revert header changes for cleaner merge.

fbergmann committed:
- expose util to bindings

fbergmann reviewed (approved):
@luciansmith thanks for this, this seems to work fine for me. I've added a bit to expose the util file to the bindings. With that i'm able to call it from python

luciansmith commented:
Whoops--exposing the util functions actually created a bunch of other errors as well in the rapid build; it's not just the ubuntu failed builds.  Re-opening this to see if it can be fixed.

luciansmith commented:
For the full build, on Windows:
2025-04-10T13:50:58.1775605Z csharp-files\libsbml.cs(531,97): error CS1503: Argument 1: cannot convert from 'System.Runtime.InteropServices.HandleRef' to 'string'

On Ubuntu:
2025-04-10T13:37:44.6406371Z /home/runner/work/libsbml/build/src/bindings/java/libsbml_wrap.cpp:3118:8: error: invalid array assignment
2025-04-10T13:37:44.6407247Z  3118 |   arg4 = *argp4;
2025-04-10T13:37:44.6407578Z       |   ~~~~~^~~~~~~~

and:

2025-04-10T13:37:44.1821060Z /home/runner/work/libsbml/build/src/bindings/csharp/libsbml_wrap.cpp:2094:8: error: invalid array assignment
2025-04-10T13:37:44.1821568Z  2094 |   arg4 = *argp4;
2025-04-10T13:37:44.1821768Z       |   ~~~~~^~~~~~~~

For the 'rapid' build, on windows:

csharp-files\libsbml.cs(531,97): error CS1503: Argument 1: cannot convert from 'System.Runtime.InteropServices.HandleRef' to 'string'

On ubuntu:
2025-04-10T13:35:02.5332898Z /home/runner/work/libsbml/build/src/bindings/r/libsbml_wrap.cpp: In function ‘SEXPREC* R_swig_c_locale_vsnprintf(SEXP, SEXP, SEXP, SEXP, SEXP)’:
2025-04-10T13:35:02.5334317Z /home/runner/work/libsbml/build/src/bindings/r/libsbml_wrap.cpp:5655:14: error: invalid array assignment
2025-04-10T13:35:02.5335180Z  5655 |         arg4 = *(reinterpret_cast< va_list * >(argp4));
2025-04-10T13:35:02.5335646Z       |         ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

On manylinux:
/__w/libsbml/build/src/bindings/java/libsbml_wrap.cpp: In function ‘jint Java_org_sbml_libsbml_libsbmlJNI_c_1locale_1vsnprintf(JNIEnv*, jclass, jstring, jlong, jstring, jlong)’:
/__w/libsbml/build/src/bindings/java/libsbml_wrap.cpp:3973:8: error: invalid array assignment
 3973 |   arg4 = *argp4;
      |   ~~~~~^~~~~~~~

luciansmith commented:
[I can't actually re-open this; addressed in other PRs instead.]
